Support O_TMPFILE and allow MOVE_BLOCKS into released extents

agrover opened the pull request:
Support O_TMPFILE: Create an unlinked file and put it on the orphan list.
If it ever gains a link, take it off the orphan list.

Change MOVE_BLOCKS ioctl to allow moving blocks into offline extent ranges.
Ioctl callers must set a new flag to enable this.

Some RH-specific struct wrapper stuff.

Add a test that tests both creating tmpfiles as well as moving their
contents into a destination file via MOVE_BLOCKS.

Signed-off-by: Andy Grover <agrover@versity.com>

zabbo commented on kmod/src/dir.c:
```diff
@@ -1760,6 +1766,42 @@ static int scoutfs_dir_open(struct inode *inode, struct file *file)
+	scoutfs_inode_index_unlock(sb, &ind_locks);
+
+	insert_inode_hash(inode);
+	d_tmpfile(dentry, inode);
```
d_tmpfile() seems to dec nlink in the vfs inode.  It needs to do this before we call update_inode_item() which copies the vfs inode to the persistent item's inode.  The general pattern is to do all our structures, then the vfs stuff, then update and unlock and release all the transactions.  So move these insert_inode_hash() and d_tmpfile() calls up to where we update the times.

zabbo commented on kmod/src/dir.c:
```diff
@@ -870,6 +870,12 @@ static int scoutfs_link(struct dentry *old_dentry,
+	/* nlink 0 -> 1 (tmpfile?) but now should no longer be */
+	if (inode->i_nlink == 1) {
+		ret = remove_orphan_item(sb, scoutfs_ino(inode));
+		WARN_ON_ONCE(ret);
```
We can handle the errors if we do this higher up.  Like we have creation and deletion wrappers in inode.c for orphan handling, we can add a item dirtying wrapper.  It'd call _item_dirty() instead of create/delete.  If you called that dirty wrapper up after dirty_inode_item then you know you have a dirty orphan in the item cache.  You could call orphan deletion after add_entry_items and know that it can't fail 'cause the item was dirty.  I'd make dirtying the orphan item and removing it conditional on some bool like "del_orphan = inode->nlink == 0" , maybe before setting up that dir_size.

agrover replied:
ok took a pass at this.

zabbo commented on kmod/src/dir.c:
```diff
@@ -1770,7 +1812,10 @@ const struct file_operations scoutfs_dir_fops = {
+
+
+const struct inode_operations_wrapper scoutfs_dir_iops = {
+	.ops = {
```
Ooof, *shakes fist at kabi*.  I guess we'll just do this and deal with the breakage and compat as we move forward to mainline.  Please describe this in the commit message.

agrover replied:
want a separate commit for this?

zabbo replied:
Nah, just describe in the commit message that we had to add nested structs for the rhel kabi expansion and that we'll run into compat pain when we get it building in mainline.

zabbo commented on kmod/src/inode.c:
```diff
@@ -1417,7 +1418,7 @@ static void init_orphan_key(struct scoutfs_key *key, u64 rid, u64 ino)
 }
 
-static int remove_orphan_item(struct super_block *sb, u64 ino)
+int remove_orphan_item(struct super_block *sb, u64 ino)
```
If it's exported it should start with scoutfs_.  To make it clear it's just a wrapper around setting a key and calling an item func, I'd have the naming match the item funcs.. so like scoutfs_orphan_delete() here.  ( and _dirty() for that dirty wrapper ).

zabbo commented on kmod/src/ioctl.h:
```diff
@@ -454,11 +457,14 @@ struct scoutfs_ioctl_alloc_detail_entry {
  * EISDIR: either the source or destination is a directory.
  * ENODATA: either the source or destination file have offline extents.
  */
+#define SCOUTFS_IOC_MB_STAGE		(1 << 0)
```
When we add flags we need to add error handling to return errors if unknown flag bits are set.  We've done it a few ways in a few places, I think I'd try _UKNOWN (U64_MAX << 1) after the _STAGE definition.  Then test flags & _UNKNOWN in the ioctl and return EINVAL.

zabbo commented on kmod/src/ioctl.h:
```diff
@@ -454,11 +457,14 @@ struct scoutfs_ioctl_alloc_detail_entry {
 	__u64 from_off;
 	__u64 len;
 	__u64 to_off;
+	__u64 flags;
```
It occurs to me that if we're adding staging we need to also check the data_version so that we can't use move_blocks to insert stale old extents into the file if a write sneaks in before us.  We'd add it to the ioctl here, and after grabbing the i_mutex in move_blocks we'd make sure that the data_version ioctl arg matches the inode data_version if we're staging, and return -ESTALE if not.  Like the release ioctl does.

zabbo commented on kmod/src/ioctl.h:
```diff
@@ -442,6 +442,9 @@ struct scoutfs_ioctl_alloc_detail_entry {
  * progress may have been made and inode fields may have been updated.
  *
+ * If MB_STAGE flag is set, destination range is allowed to be in an
+ * offline extent.
```
We should describe that the destination range *has* to be an offline extent, what we'll return if it's not, and we should add a description of requiring data_version as well when staging.

zabbo reviewed (changes requested):
Nothing scary, just some light refactoring and adding the data_version and I think it looks good!

agrover commented:
* addressed comments
* also fixed some other UNKNOWN flag bitmasks that weren't wide enough.

TODO
* document kabi ugliness, maybe in a separate commit

versity-github commented:
Build 50 http://jenkins-prod.vpn.versity.com:8080/job/scoutfs-run-pr/50/ git: 557855878a963665f1bc8025632603207ed8ffec

zabbo commented on kmod/src/inode.c:
```diff
@@ -1417,7 +1418,23 @@ static void init_orphan_key(struct scoutfs_key *key, u64 rid, u64 ino)
+	init_orphan_key(&key, sbi->rid, ino);
+
+	ret = scoutfs_item_dirty(sb, &key, lock);
+	if (ret == -ENOENT)
```
Ah, that's why it wasn't returning -ENOENT.  That's not OK :).   It really needs to dirty or return an error.

agrover replied:
oooooooooops!

zabbo commented on kmod/src/dir.c:
```diff
@@ -855,6 +858,10 @@ static int scoutfs_link(struct dentry *old_dentry,
 	if (ret)
 		goto out;
 
+	ret = scoutfs_orphan_dirty(sb, scoutfs_ino(inode));
```
We only want to dirty if we're deleting the orphan.. was this not returning -ENOENT for links when we didn't have the orphan item yet?

zabbo commented on kmod/src/ioctl.h:
```diff
@@ -435,30 +436,41 @@ struct scoutfs_ioctl_alloc_detail_entry {
 	__u64 from_off;
 	__u64 len;
 	__u64 to_off;
+	__u64 data_version;
```
I went and looked and the mb in the move_blocks command in utils isn't zeroing the arg struct.. so we need to either zero or explicitly initialize the new flags/dv fields.

zabbo commented on kmod/src/data.c:
```diff
@@ -1274,10 +1281,31 @@ int scoutfs_data_move_blocks(struct inode *from, u64 from_off,
+			}
 			if (ret < 0)
 				break;
 
```
(closest line I can comment on, alsjdfkjsafd.)  We still need to turn the err = ext_remove below into using _set to revert back to the offline extent when is_stage, otherwise use the existing _remove undoing for the non-stage case.  They can both still bug_on if they fail :/.

agrover replied:
ok cool.

zabbo reviewed (changes requested):
Lookin' good!  A few more tweaks and we're gold.

versity-github commented:
Build 55 http://jenkins-prod.vpn.versity.com:8080/job/scoutfs-run-pr/55/ git: 375ee01b7c723999af2f7aa3f4305f81e6226599

versity-github commented:
Build 56 http://jenkins-prod.vpn.versity.com:8080/job/scoutfs-run-pr/56/ git: 98360dce16df5378801d04341f8f9cf122dd3444

zabbo commented on kmod/src/data.c:
```diff
@@ -1274,21 +1281,47 @@ int scoutfs_data_move_blocks(struct inode *from, u64 from_off,
+					/* re-mark dest range as offline */
+					err = scoutfs_ext_set(sb, &data_ext_ops, &to_args,
+							      to_start, len,
+							      0, ext.flags | SEF_OFFLINE);
```
I don't think we should assume that ext.flags | offline == to_ext.flags.  That might not even be the case today if we try to move unwritten extents, and certainly might not be the case if we add weird flags later.  Can we move the to_ext from being local to its block so that we can use to_ext.flags here to restore the original flags?    (maybe call it off_ext?)

agrover replied:
ah yes that's the right thing to do.

zabbo reviewed (changes requested):
Just this one quick flag maintenance fix and it looks good!

versity-github commented:
Build 57 http://jenkins-prod.vpn.versity.com:8080/job/scoutfs-run-pr/57/ git: 0deb232d3f15676b461fd73191cca2829c78cc26
